### PR TITLE
[trusted state] fix the ratchet when epoch_proof has more

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation

The ratchet would fail when the latest_li is in the future epoch and epoch_change_proof doesn't forward that far and indicates more=true.

This commit fixes the issue by fallback to the epoch_change_li when the epoch_change_proof doesn't result in the same epoch as latest_li.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

Added test.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: types/src/trusted_state.rs [hunk 1]
```diff
@@ -94,17 +94,27 @@ impl TrustedState {
...
+                    latest_li.ledger_info()
+                } else {
+                    ensure!(
+                        latest_li == epoch_change_li
+                            || latest_li.ledger_info().epoch() > new_epoch
+                                && epoch_change_proof.more,
+                        "Invalid epoch in latest ledger info",
+                    );
+                    epoch_change_li.ledger_info()
+                });
```
Comment (davidiw) [inline types/src/trusted_state.rs:114]: this isn't really readable.... the level of nesting here is :trollface:

Comment (davidiw) [reply types/src/trusted_state.rs:114]: after reading this a dozen times I can see the logic is correct, but we should really break it up with a clear comment that you're moving into a partial state as you attempt to obtain a new set of li.

There are actually three states here...

the new li doesn't move us into a new epoch, okay, let' s just use it
the new li is in the new epoch, so consume the epoch change li
the new li is beyond the new epoch but there are more epochs to come, so consume the epoch change li
then we have the one error case, which is new li is beyond the the new epoch but we don't have more epochs coming, error

can we write this as a if else if else if else  that returns the appropriate li?

**Comment by davidiw** [comment]
amazing find, now do something with this in safety rules :P

File: types/src/trusted_state.rs (hunk 1 above)
Comment (phlip9) [reply types/src/trusted_state.rs:114]: It looks like the key change here is `latest_li.ledger_info().epoch() > new_epoch && epoch_change_proof.more`, i.e., there was a gap and you need more epoch change li's to catch up, so just ratchet as far as you can. Actually, this makes me wonder why we even have `more`? Could just be an implicit property like "there is a gap" ==> "more".

Reading through how `libra-client` uses this also reminded me how we're not checking other proofs lol. 😭

Comment (zekun000) [reply types/src/trusted_state.rs]: agree `more` is more like a hint. I don't know what's our story of verifying-api overall.

**Review by phlip9 — APPROVED** [review]

**Pull request closed without merging**